### PR TITLE
Cleanup and redundant code removal, better setup debugging

### PR DESCRIPTION
**Pull request by brontide**
_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

**By submitting this pull request, I confirm the following (please check boxes, eg [X]):**
- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/pi-hole/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:**
- [X] 7
---
### Visible changes
- Lines that are warnings start with "***" rather than ":::" to draw attention
- If any warnings are encountered then the installer stops at the end and prompts to continue.
- PIHOLE_DEPS have been moved to pre-install/update per @PromoFaux 
- The nice thing is for uncomplicated installs this should be little to no difference
### Other changes
- Removed a lot of duplication from install and update merging it back into main.
- Removed some unneeded code

I have run it against an intentionally broken selinux install to show the results.  I will run it past my collection of 10 containers to see if there any new issues.

```
[root@pihole-fedora24 ~]# ./basic-install.sh 
:::
::: You are root.
::: Verifying free disk space...
:::
::: Checking dnf for upgraded packages.... done!
:::
::: Your system is up to date! Continuing with Pi-hole installation...
:::    Checking for iproute... installed!
:::    Checking for net-tools... installed!

::: IPv4 address: 10.195.42.144/24
::: IPv6 address: 
::: Using Google DNS servers.
::: Checking if user 'pihole' exists...
::: User 'pihole' doesn't exist. Creating...

*** Warning: 'lighty-enable-mod' utility not found. Please ensure fastcgi is enabled if you experience issues.
:::
::: Installing scripts to /opt/pihole... done.
:::
::: Installing configs...
:::    Existing dnsmasq.conf found... it is not a pi-hole file, leaving alone!
:::    Copying 01-pihole.conf to /etc/dnsmasq.d/01-pihole.conf... done.
:::
::: Creating log file and changing owner to dnsmasq... done!
:::
::: Installing pihole custom index page...
::: No default index.lighttpd.html file found... not backing up done!
::: Installing sudoer file... done!
:::
::: Preparing to run gravity.sh to refresh hosts...
::: Running gravity.sh
:::
::: Neutrino emissions detected...
:::
::: No custom adlist file detected, reading from default file... done!
:::
::: Existing pihole directory found
:::
::: Getting raw.githubusercontent.com list... List updated, transport successful!
::: Getting mirror1.malwaredomains.com list... List updated, transport successful!

::: Cleaning up un-needed files... done!
:::
::: Refresh lists in dnsmasq...:::
::: Installing latest Cron script... done!
::: Configuring iptables for httpd and dnsmasq..

::: SELinux Detected
::: Checking for SELinux policy development packages... installed!
::: Enabling httpd server side includes (SSI).. SELinux not enabled

::: Compiling Pi-Hole SELinux policy..
checkmodule:  loading policy configuration from /etc/.pihole/advanced/selinux/pihole.te
checkmodule:  policy configuration loaded
checkmodule:  writing binary representation (version 17) to /etc/pihole/pihole.mod
Failed to resolve 'cil_gen_require' in typeattributeset statement at /var/lib/selinux/targeted/tmp/modules/400/pihole/cil:1
semodule:  Failed!
*** Warning: Pi-Hole SELinux policy did not install.
::: Restarting services...
:::
::: Starting dnsmasq service... done.
:::
::: Enabling dnsmasq service to start on reboot... done.
:::
::: Starting lighttpd service... done.
:::
::: Enabling lighttpd service to start on reboot... done.
::: done.
:::
::: Installation Complete! Configure your devices to use the Pi-hole as their DNS server using:
:::     10.195.42.144
:::     
:::
::: If you set a new IP address, you should restart the Pi.
:::
::: The install log is located at: /etc/pihole/install.log
::: View the web interface at http://pi.hole/admin or http://10.195.42.144/admin
*** ***************************************************************************
*** INSTALLER ENCOUNTERED ONE OR MORE BUGS, PLEASE REVIEW LOGS AND CHECK INSTALLATION
*** PLEASE PRESS ENTER TO CONFIRM 
```

### CODE REVIEW

File: automated install/basic-install.sh [hunk 1]
```diff
@@ -979,12 +970,27 @@ main() {
...
+	    install_dependent_packages PIHOLE_DEPS[@]
 		# Install and log everything to a file
 		installPihole | tee ${tmpLog}
 	else
+	    # Install packages used by the Pi-hole
+	    install_dependent_packages PIHOLE_DEPS[@]
 		updatePihole | tee ${tmpLog}
 	fi
 
+    (
```
Comment (dschaper) [inline automated install/basic-install.sh]: Clever! Will steal this idea :)

**Comment by dschaper** [comment]
Look forward to the battery of test results..

**Comment by brontide** [comment]
pihole-alpine34
---

```
git checkout misc_fixed
git merge origin/alpine_support
```

Fix a few minor merge issues

lighttpd does not start, don't know why... dns functions are fine.

Looks like a regression in one of the other scripts.

``` 
grep: /etc/pihole/blacklist.txt: No such file or directory
grep: /etc/pihole/blacklist.txt: No such file or directory
grep: /etc/pihole/blacklist.txt: No such file or directory
grep: /etc/pihole/blacklist.txt: No such file or directory
grep: /etc/pihole/blacklist.txt: No such file or directory
grep: /etc/pihole/blacklist.txt: No such file or directory
grep: /etc/pihole/blacklist.txt: No such file or directory
grep: /etc/pihole/blacklist.txt: No such file or directory
```

pihole-centos6
---

Did not install, procps-ng

pihole-centos7
---

Stopped on php, restarted was fine.  Created directory

```
install -d -o lighttpd -g lighttpd -m 770 /var/cache/lighttpd/uploads
```

Full function

pihole-debian-sid
---

Install stopped after dnsmasq, restarted the network and it completed.

Full function

pihole-debian7
---

```
git checkout misc_fixes
git merge origin/wheezy_fix  #no conflicts
```

Fully functional

pihole-debian8
---

Installed and fully functional in one go.

pihole-fedora24
---

Installed and fully functional in one pass

pihole-u1404
---

Fully functional in one pass

pihole-u1604
---

Fully functional in one pass

pihole-u1610
---

Fully function in one pass.

**Comment by brontide** [comment]
It's ready for review and merge.

File: automated install/basic-install.sh [hunk 2]
```diff
@@ -1048,6 +1051,12 @@ main() {
 	echo ":::"
 	echo "::: The install log is located at: /etc/pihole/install.log"
 	echo "::: View the web interface at http://pi.hole/admin or http://${IPv4_address%/*}/admin"
+
+    [ ${INSTALLBUG} ] && (
+        echo "*** ***************************************************************************"
+        echo "*** INSTALLER ENCOUNTERED ONE OR MORE BUGS, PLEASE REVIEW LOGS AND CHECK INSTALLATION"
+        read -p "*** PLEASE PRESS ENTER TO CONFIRM "
```
Comment (PromoFaux) [inline automated install/basic-install.sh]: If the script is `curl`'d  The script will fall out here anyway. Found that out a while back! Curling to bash doesn't support reading user input.

File: automated install/basic-install.sh (hunk 1 above)
Comment (PromoFaux) [inline automated install/basic-install.sh]: You can put in a collection with your `}`

**Comment by PromoFaux** [comment]
```
pi@raspberrypi:/etc/.pihole $ pihole -r
:::
::: You are root.
::: Repair option selected.
::: Verifying free disk space...
:::
::: apt-get update has not been run today. Running now... done!
:::
::: Checking apt-get for upgraded packages.... done!
:::
::: There are 22 updates available for your system!
::: We recommend you run 'apt-get upgrade' after installing Pi-Hole!
:::
:::    Checking for apt-utils... installed!
:::    Checking for whiptail... installed!
:::    Checking for git... installed!
:::    Checking for dhcpcd5... installed!
::: --reconfigure passed to install script. Not downloading/updating local repos
:::    Checking for dnsutils... installed!
:::    Checking for bc... installed!
:::    Checking for dnsmasq... installed!
:::    Checking for lighttpd... installed!
:::    Checking for php5-common... installed!
:::    Checking for php5-cgi... installed!
:::    Checking for curl... installed!
:::    Checking for unzip... installed!
:::    Checking for wget... installed!
:::    Checking for sudo... installed!
:::    Checking for netcat... installed!
:::    Checking for cron... installed!
:::    Checking for iproute2... installed!
:::
::: Installing scripts to /opt/pihole... done.
:::
::: Installing configs...
:::    Existing dnsmasq.conf found... it is not a pi-hole file, leaving alone!
:::    Copying 01-pihole.conf to /etc/dnsmasq.d/01-pihole.conf... done.
:::
::: Creating log file and changing owner to dnsmasq... already exists!
:::
::: Installing pihole custom index page... Existing page detected, not overwriting
::: Installing sudoer file... done!
:::
::: Preparing to run gravity.sh to refresh hosts...
::: Cleaning up previous install (preserving whitelist/blacklist)
::: Running gravity.sh
:::
::: Neutrino emissions detected...
:::
::: No custom adlist file detected, reading from default file... done!
:::
::: Getting raw.githubusercontent.com list... List updated, transport successful!
::: Getting mirror1.malwaredomains.com list... List updated, transport successful!
::: Getting sysctl.org list... List updated, transport successful!
::: Getting zeustracker.abuse.ch list... List updated, transport successful!
::: Getting s3.amazonaws.com list... List updated, transport successful!
::: Getting s3.amazonaws.com list... List updated, transport successful!
::: Getting hosts-file.net list... List updated, transport successful!
::: Getting raw.githubusercontent.com list... List updated, transport successful!
:::
::: Aggregating list of domains... done!
::: Formatting list of domains to remove comments.... done!
::: 125647 domains being pulled in by gravity...
::: Removing duplicate domains.... done!
::: 101569 unique domains trapped in the event horizon.
:::
::: Adding adlist sources to the whitelist... done!
::: Whitelisting 9 domains... done!
::: BlackListing 4 domains... done!
::: Formatting domains into a HOSTS file...
:::
::: Cleaning up un-needed files... done!
:::
::: Refresh lists in dnsmasq...
::: Pi-hole blocking is Enabled
:::
::: Installing latest Cron script... done!
::: Configuring iptables for httpd and dnsmasq..
::: Restarting services...
:::
::: Starting dnsmasq service... done.
:::
::: Enabling dnsmasq service to start on reboot... done.
:::
::: Starting lighttpd service... done.
:::
::: Enabling lighttpd service to start on reboot... done.
::: done.
:::
::: Update complete!
:::
::: The install log is located at: /etc/pihole/install.log
::: View the web interface at http://pi.hole/admin or http:///admin
*** ***************************************************************************
*** INSTALLER ENCOUNTERED ONE OR MORE BUGS, PLEASE REVIEW LOGS AND CHECK INSTALLATION
*** PLEASE PRESS ENTER TO CONFIRM
```

Install output does not give any indication as to what went wrong here. 

How about setting `INSTALLBUG` to a particular in each place that it is set, so that we have an error code and easily identify what went wrong. If it's possible to hit `INSTALLBUG` more than once, then perhaps it can be an array?

**Comment by PromoFaux** [comment]
Also can we move [this section](https://github.com/brontide/pi-hole/blob/36e2719cf0adae5c8c83dfda9ca2316913bdb14f/automated%20install/basic-install.sh#L978-L984) to after the dependency check? Looks a bit weird in the middle (see above comment)

**Comment by dschaper** [comment]
Started the array idea in #877, very basic to start but it does even work on a system without the Pi-hole even installed.

File: automated install/basic-install.sh (hunk 2 above)
Comment (brontide) [inline automated install/basic-install.sh]: It's the last line anyways, but I'll see if I can clear that.

**Pull request closed without merging**